### PR TITLE
Update RemoteFile::local

### PR DESCRIPTION
**Pull request by iarspider**
#### PR description:

* Got rid of `Wunused-result` (from call to `c_str()`)
* Since C-string `&(temp[0])` is not guaranteed by C++ standard to be null-terminated (but, I think, GCC's implementation kind of guarantees this), copy string template into a vector, explicitly add `\0` and pass it to `mkstemp`. Only `c_str()` is guaranteed to return null-terminated string, but writing to the character array accessed through `c_str()` is undefined behavior.
* Replaced call to `getcwd` with C++17 filesystem API

#### PR validation:

Bot tests

### CODE REVIEW

**Comment by iarspider** [comment]
please test

File: Utilities/StorageFactory/src/RemoteFile.cc [hunk 1]
```diff
@@ -53,27 +54,27 @@ int RemoteFile::local(const std::string &tmpdir, std::string &temp) {
...
-    char *buf = (char *)malloc(len);
-    getcwd(buf, len);
-
-    temp.reserve(len + 32);
-    temp = buf;
-    free(buf);
+    fs::path current_path = fs::current_path();
+    temp = current_path.string();
   } else {
-    temp.reserve(tmpdir.size() + 32);
```
Comment (Dr15Jones) [inline Utilities/StorageFactory/src/RemoteFile.cc:64]: I think using temp here and then temp_chars is not a good way to go. My suggestion is to just declare `temp_chars` early and copy all the info from `current_path` and `storage-factory-local-XXXXXX` directly into `temp_chars`, use `temp_chars` in `mkstemp` and only when done, copy `temp_chars` to `temp`.

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

File: Utilities/StorageFactory/src/RemoteFile.cc [hunk 2]
```diff
@@ -52,27 +54,30 @@ int RemoteFile::local(const std::string &tmpdir, std::string &temp) {
...
+  temp_chars.insert(temp_chars.end(), suffix.begin(), suffix.end());
 
-  int fd = mkstemp(&temp[0]);
+  // Ensure the vector is null-terminated
+  temp_chars.push_back('\0');
+
+  int fd = mkstemp(temp_chars.data());
   if (fd == -1)
-    throwStorageError("RemoteFile", "Calling RemoteFile::local()", "mkstemp()", errno);
+    throw std::system_error(errno, std::generic_category(), "mkstemp failed");
```
Comment (Dr15Jones) [inline Utilities/StorageFactory/src/RemoteFile.cc:77]: Please use a `throwStorageError` as that used a cms specific exception type and adds necessary additional information to help diagnose problems.

File: Utilities/StorageFactory/src/RemoteFile.cc [hunk 3]
```diff
@@ -52,27 +54,30 @@ int RemoteFile::local(const std::string &tmpdir, std::string &temp) {
...
-    temp.reserve(len + 32);
-    temp = buf;
-    free(buf);
+  if (tmpdir.empty() || tmpdir == ".") {
+    std::filesystem::path current_path = std::filesystem::current_path();
+    temp_chars.assign(current_path.string().begin(), current_path.string().end());
   } else {
-    temp.reserve(tmpdir.size() + 32);
-    temp = tmpdir;
+    temp_chars.assign(tmpdir.begin(), tmpdir.end());
```
Comment (Dr15Jones) [inline Utilities/StorageFactory/src/RemoteFile.cc:63]: Please do `temp_chars.reserve(temdir.size()+30)` before the `assign` so only one allocation has to be done.

File: Utilities/StorageFactory/src/RemoteFile.cc [hunk 4]
```diff
@@ -52,27 +54,30 @@ int RemoteFile::local(const std::string &tmpdir, std::string &temp) {
...
-    char *buf = (char *)malloc(len);
-    getcwd(buf, len);
+  std::vector<char> temp_chars;
 
-    temp.reserve(len + 32);
-    temp = buf;
-    free(buf);
+  if (tmpdir.empty() || tmpdir == ".") {
+    std::filesystem::path current_path = std::filesystem::current_path();
+    temp_chars.assign(current_path.string().begin(), current_path.string().end());
```
Comment (Dr15Jones) [inline Utilities/StorageFactory/src/RemoteFile.cc:61]: The returned from `current_path.string()` is a temporary so calling it twice gives two different strings which means the begin/end are not proper pairs. Please do

```suggestion
    auto spath = std::filesystem::current_path().string();
    temp_chars.reserve(spath.size()+30);
    temp_chars.assign(spath.begin(), spath.end());
```

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

**Comment by makortel** [comment]
Comparison differences are in 29634.911, which is a "known" instability (although we don't seem to have a dedicated issue for that?)

**Comment by makortel** [comment]
+core

**Comment by rappoccio** [comment]
+1

**Pull request merged**